<a href="https://colab.research.google.com/github/VikiTarasova/DTA_2026/blob/main/Homework/task28_pandas_tasks_beginners_DTA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Практичні завдання з pandas для початківців

**Мета:** Набути базових навичок роботи з бібліотекою pandas для аналізу табличних даних.

**Інструкції:**
- Виконуйте завдання послідовно — кожне наступне спирається на попередні.
- Після виконання кожного завдання напишіть короткий висновок у комірці `**Висновок:**`.
- Не змінюйте код у комірках з підготовкою даних.

**Набір даних:** Умовна таблиця продажів інтернет-магазину за 2023 рік.

---
## Підготовка: Генерація тестових даних

Запустіть цю комірку перед початком роботи. Не змінюйте код.

In [1]:
import pandas as pd
import numpy as np

np.random.seed(42)
n = 200

categories = ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт']
cities = ['Київ', 'Харків', 'Одеса', 'Дніпро', 'Львів']
payment_methods = ['Картка', 'Готівка', 'Онлайн']

df = pd.DataFrame({
    'order_id':        range(1001, 1001 + n),
    'date':            pd.date_range('2023-01-01', periods=n, freq='D')[:n],
    'category':        np.random.choice(categories, n),
    'product':         ['Товар_' + str(i) for i in np.random.randint(1, 51, n)],
    'city':            np.random.choice(cities, n),
    'quantity':        np.random.randint(1, 10, n),
    'price':           np.round(np.random.uniform(50, 5000, n), 2),
    'payment_method':  np.random.choice(payment_methods, n),
    'rating':          np.random.choice([1, 2, 3, 4, 5, np.nan], n, p=[0.05, 0.1, 0.15, 0.3, 0.3, 0.1]),
    'is_returned':     np.random.choice([True, False], n, p=[0.1, 0.9]),
})

# Додаємо кілька пропущених значень
df.loc[np.random.choice(df.index, 10, replace=False), 'city'] = np.nan
df.loc[np.random.choice(df.index, 5, replace=False), 'price'] = np.nan

# Розраховуємо загальну суму
df['total'] = df['quantity'] * df['price']

print('Дані успішно згенеровано. Рядків:', len(df))
print('Колонки:', list(df.columns))

Дані успішно згенеровано. Рядків: 200
Колонки: ['order_id', 'date', 'category', 'product', 'city', 'quantity', 'price', 'payment_method', 'rating', 'is_returned', 'total']


---
## Завдання 1 (Легке). Перший погляд на дані

Виведіть перші 10 рядків датафрейму `df` та отримайте загальну інформацію про нього (кількість рядків, колонок, типи даних) за допомогою методів `head()` та `info()`.

In [50]:
# Ваш код тут

print(df.head(10))

df.info()

   order_id       date  category   product    city  quantity    price  \
0      1001 2023-01-01  Продукти  Товар_34    Київ         1   963.38   
1      1002 2023-01-02     Спорт   Товар_6   Львів         6  1086.28   
2      1003 2023-01-03     Книги  Товар_22  Дніпро         4  1883.84   
3      1004 2023-01-04     Спорт  Товар_11  Дніпро         1  2448.39   
4      1005 2023-01-05     Спорт  Товар_48  Дніпро         7  3110.36   
5      1006 2023-01-06      Одяг  Товар_16   Львів         9  1876.12   
6      1007 2023-01-07     Книги  Товар_33  Дніпро         4  2339.55   
7      1008 2023-01-08     Книги   Товар_9   Львів         4  3749.98   
8      1009 2023-01-09     Книги   Товар_6  Дніпро         6   231.58   
9      1010 2023-01-10     Спорт  Товар_16   Одеса         3  1299.56   

  payment_method  rating  is_returned     total   rank  month price_category  \
0         Картка     5.0        False    963.38  188.0      1        Середня   
1         Картка     4.0        Fals

**Висновок:** *(Напишіть, скільки рядків і колонок у датафреймі, які типи даних зустрічаються, чи є пропущені значення)*

Датафрейм містить 200 рядків та 11 колонок. У ньому присутні такі типи даних:  

`int64`,   
`float64`,   
`object`,   
`datetime64[ns]`,   
`bool`.  

У датафреймі є пропущені значення:  

у колонці `city` — 10 пропусків;  
у колонці `price` — 5 пропусків;  
у колонці `rating` — 23 пропусків;  
у колонці  `total` — 5 пропусків.  

Тому перед подальшим аналізом даних варто перевірити та за потреби обробити пропущені значення.  

---
## Завдання 2 (Легке). Описова статистика

Використайте метод `describe()` для отримання описової статистики числових колонок. Окремо виведіть мінімум, максимум і середнє значення колонки `price`.

In [19]:
# Ваш код тут

print("Мінімум price:", df['price'].min())
print("Максимум price:", df['price'].max())
print("Середнє price:", df['price'].mean())

print("Медіана price:", df['price'].median())

df.describe()

Мінімум price: 139.65
Максимум price: 4950.3
Середнє price: 2500.579641025641
Медіана price: 2517.29


,order_id,date,quantity,price,rating,total
count,200.000000,200,200.000000,195.000000,177.000000,195.000000
mean,1100.500000,2023-04-10 12:00:00,4.700000,2500.579641,3.813559,11629.097590
min,1001.000000,2023-01-01 00:00:00,1.000000,139.650000,1.000000,149.350000
25%,1050.750000,2023-02-19 18:00:00,2.000000,1288.215000,3.000000,3893.955000
50%,1100.500000,2023-04-10 12:00:00,5.000000,2517.290000,4.000000,8746.840000
75%,1150.250000,2023-05-30 06:00:00,7.000000,3706.975000,5.000000,16922.420000
max,1200.000000,2023-07-19 00:00:00,9.000000,4950.300000,5.000000,44324.640000
std,57.879185,NaN,2.618066,1380.066748,1.169808,9968.293697


**Висновок:** *(Опишіть діапазон цін, середнє значення, чи є суттєвий розкид)*  
Ціни в даному наборі даних варіюються від 139.65 до 4950.30, що свідчить про широкий діапазон значень. Середня ціна становить приблизно 2500.58, тобто більшість товарів знаходяться в середньому ціновому сегменті.  

Медіана (50%) дорівнює 2517.29, що дуже близько до середнього значення. Це означає, що розподіл цін є відносно симетричним, без сильного перекосу в один бік.  

Стандартне відхилення становить приблизно 1380, що є доволі високим значенням. Це вказує на суттєвий розкид цін — товари сильно відрізняються за вартістю (є як дешеві, так і дорогі позиції).  

**Підсумок:**  
- Мінімальна ціна: 139.65  
- Максимальна ціна: 4950.30  
- Середня ціна: ~2500.58  
- Розкид: значний (висока варіативність цін)  

---
## Завдання 3 (Легке). Пропущені значення

Знайдіть кількість пропущених значень у кожній колонці за допомогою `isnull().sum()`. Визначте, який відсоток рядків має пропущені значення хоча б в одній колонці (`dropna()` та `len()`).

In [24]:
# Ваш код тут
df.isnull().sum()



,0
order_id,0
date,0
category,0
product,0
city,10
quantity,0
price,5
payment_method,0
rating,23
is_returned,0


In [26]:
total_cells = df.size
missing_cells = df.isnull().sum().sum()

print("Всього ячейок:", total_cells)
print("Пропущених ячейок:", missing_cells)
print("Відсоток пропусків:", missing_cells / total_cells * 100)

Всього ячейок: 2200
Пропущених ячейок: 43
Відсоток пропусків: 1.9545454545454546


**Висновок:** *(У яких колонках є пропуски, наскільки це критично для аналізу)*  
Пропущені значення розподілені по окремих колонках і становлять невелику частку від загальної кількості ячейок у датасеті. Загальний відсоток пропусків є низьким, тому дані в цілому придатні для аналізу, але потребують часткової обробки пропущених значень у колонках `city`, `price` та `rating`.

---
## Завдання 4 (Легке). Відбір колонок і рядків

Створіть новий датафрейм `df_subset`, що містить лише колонки `order_id`, `category`, `city`, `total`. Потім виведіть рядки з індексами від 10 до 20 включно за допомогою `.iloc[]`.

In [28]:
# Ваш код тут
df_subset = df[['order_id', 'category', 'city', 'total']]

df_subset.iloc[10:21]

,order_id,category,city,total
10,1011,Продукти,Дніпро,21486.48
11,1012,Книги,Львів,31368.89
12,1013,Спорт,Харків,7748.40
13,1014,Одяг,Дніпро,18787.72
14,1015,Продукти,NaN,1741.50
15,1016,Одяг,Одеса,4529.38
16,1017,Продукти,Київ,10745.84
17,1018,Спорт,Одеса,10001.84
18,1019,Електроніка,Дніпро,12444.75
19,1020,Продукти,Харків,13422.92


**Висновок:** *(Поясніть різницю між вибором колонок через список і через `iloc`)*  
Вибір колонок через список `(df[['order_id', 'category', 'city', 'total']])` дозволяє звертатися до даних за назвами колонок, тобто явно вказувати, які ознаки потрібні. Це більш читабельний і безпечний спосіб, оскільки не залежить від порядку колонок у датафреймі.  

Метод `.iloc[]` використовується для вибору даних за позицією (індексами). Наприклад, `df_subset.iloc[10:21]` означає вибір рядків з 10 по 20 включно за їх порядковим номером, незалежно від назв.  

Отже:  

вибір через список — працює з іменами колонок;  
`.iloc[]` — працює з числовими індексами (позиціями).

---
## Завдання 5 (Легке). Фільтрація даних

Відфільтруйте датафрейм і отримайте:
1. Всі замовлення з категорії `'Електроніка'`.
2. Всі замовлення, де `price` перевищує 2000.
3. Поверненi замовлення (`is_returned == True`) з міста `'Київ'`.

Виведіть кількість рядків для кожного фільтру.

In [36]:
# Ваш код тут
electronics = df[df['category'] == 'Електроніка']
price_gt_2000 = df[(df['category'] == 'Електроніка') & (df['price'] > 2000)]
kyiv_returns = df[
    (df['category'] == 'Електроніка') &
    (df['city'] == 'Київ') &
    (df['is_returned'] == True)
]
kyiv_orders = df[df['city'] == 'Київ']
return_rate = len(kyiv_returns) / len(kyiv_orders) * 100


print("Електроніка:", len(electronics))
print("Електроніка Price > 2000:", len(price_gt_2000))
print("Повернення з Києва:", len(kyiv_returns))
print("Повернення з Києва (%):", return_rate)

Електроніка: 43
Електроніка Price > 2000: 23
Повернення з Києва: 2
Повернення з Києва (%): 5.405405405405405


**Висновок:** *(Яка частка замовлень відповідає кожному критерію? Що можна сказати про повернення з Києва?)*  

Аналіз даних по категорії «Електроніка» показує, що вона є важливою частиною продажів — всього 43 замовлення, з яких 23 належать до дорогого сегмента (price > 2000). Це означає, що більше половини покупок припадає на високовартісні товари, що формує сильний преміум-попит.  

Рівень повернень у Києві становить лише 5.4% (2 повернення), що є низьким показником і свідчить про достатньо високу якість товарів та задоволеність клієнтів у цьому регіоні.  

Рекомендація: варто додатково проаналізувати, товари з яким ціновим діапазоном найчастіше повертаються, щоб виявити ризикові сегменти (наприклад, низький, середній або високий ціновий рівень).  


---
## Завдання 6 (Середнє). Сортування та ранжування

1. Відсортуйте датафрейм за колонкою `total` у спадному порядку та виведіть топ-10 найбільших замовлень.
2. Додайте нову колонку `rank`, що містить порядковий номер замовлення за розміром суми (від найбільшого до найменшого) за допомогою `.rank()`.

In [40]:
# Ваш код тут
top_10 = df.sort_values(by='total', ascending=False).head(10)
print(top_10)

df['rank'] = df['total'].rank(ascending=False, method='dense')


     order_id       date     category   product    city  quantity    price  \
190      1191 2023-07-10         Одяг  Товар_22    Київ         9  4924.96   
58       1059 2023-02-28     Продукти  Товар_33  Харків         9  4848.88   
186      1187 2023-07-06        Спорт  Товар_34    Київ         9  4675.12   
55       1056 2023-02-25  Електроніка   Товар_3  Дніпро         8  4912.77   
169      1170 2023-06-19     Продукти  Товар_50    Київ         8  4739.05   
191      1192 2023-07-11        Спорт  Товар_25  Харків         9  4202.55   
193      1194 2023-07-13         Одяг  Товар_22     NaN         8  4608.17   
51       1052 2023-02-21     Продукти  Товар_35  Дніпро         8  4488.99   
25       1026 2023-01-26        Книги  Товар_18  Дніпро         8  4458.11   
143      1144 2023-05-24     Продукти  Товар_31    Київ         9  3732.53   

    payment_method  rating  is_returned     total  rank  
190        Готівка     4.0         True  44324.64   1.0  
58          Картка     5.

**Висновок:** *(Яка максимальна сума замовлення? З яких категорій найбільші замовлення?)*  
Максимальна сума замовлення становить 44 324.64, але це замовлення було повернене, отже максимальна сума замовлення становить - 43 639.92.  

Топ-10 найбільших замовлень формуються переважно з категорій «Продукти», «Спорт», «Одяг» та «Електроніка», тобто великі чеки розподілені між кількома сегментами, а не концентруються в одній категорії.

---
## Завдання 7 (Середнє). Групування та агрегація

За допомогою `groupby()` розрахуйте для кожної категорії товарів:
- загальний обсяг продажів (сума `total`),
- кількість замовлень,
- середню ціну (`price`).

Відсортуйте результат за загальним обсягом продажів.

In [44]:
# Ваш код тут
result = df.groupby('category').agg(
    total_sales=('total', 'sum'),
    orders_count=('order_id', 'count'),
    avg_price=('price', 'mean')
).sort_values(by='total_sales', ascending=False)

print(result)

             total_sales  orders_count    avg_price
category                                           
Продукти       583761.68            49  2513.979787
Одяг           474135.61            35  2707.906176
Електроніка    428129.79            43  2334.470698
Спорт          400802.56            36  2494.205143
Книги          380844.39            37  2491.881944


**Висновок:** *(Яка категорія приносить найбільший дохід? Чи збігається це з кількістю замовлень?)*  

Найбільший дохід приносить категорія «Продукти» — 583 761.68, що робить її ключовим драйвером виручки в датасеті. Далі йдуть «Одяг» (474 135.61) та «Електроніка» (428 129.79).  

При цьому найбільша кількість замовлень не повністю збігається з лідером по доходу:  

- «Продукти» мають і найбільший дохід, і найбільшу кількість замовлень (49) → це головний сегмент,  
- «Електроніка» має багато замовлень (43), але менший дохід,
- «Одяг» має менше замовлень (35), але високий середній чек, тому займає друге місце по доходу.  

**Рекомендації**:  
- Сфокусуватися на категорії «Продукти», оскільки вона приносить найбільший дохід і має найбільшу кількість замовлень — це стабільний основний драйвер виручки.  
- Посилити монетизацію «Електроніки», адже при великій кількості замовлень її дохід нижчий — варто перевірити середній чек, структуру товарів і можливості апселу (дорожчі комплекти, допродажі).  
- Підтримувати «Одяг» як високочековий сегмент, оскільки він має менше замовлень, але високий дохід — потенціал росту тут у збільшенні кількості продажів.  
- Аналізувати цінову стратегію «Спорту» і «Книг», щоб зрозуміти, як збільшити їхній вклад у загальну виручку (акції, бандли, знижки).



---
## Завдання 8 (Середнє). Обробка пропущених значень

1. Заповніть пропущені значення в колонці `price` медіаною цін відповідної категорії (використайте `groupby().transform('median')`).
2. Заповніть пропущені значення в колонці `city` рядком `'Невідомо'`.
3. Пропущені значення в `rating` залиште без змін — поясніть чому у висновку.

Перевірте, що пропусків у `price` та `city` більше немає.

In [47]:
# Ваш код тут
df['price'] = df['price'].fillna(
    df.groupby('category')['price'].transform('median')
)

df['city'] = df['city'].fillna('Невідомо')
print(df[['price', 'city', 'rating']].isnull().sum())

price      0
city       0
rating    23
dtype: int64


**Висновок:** *(Чому медіана краща за середнє для заповнення цін? Чому пропуски в rating можна залишити?)*  

Медіана менш чутлива до дуже високих або дуже низьких значень. У наборі даних є дорогі замовлення, які можуть суттєво зміщувати середнє значення, тому медіана краще відображає типову ціну товару в категорії.  

Пропуски в колонці `rating` доцільно залишити без змін, оскільки відсутність рейтингу не означає певну оцінку. Найчастіше це свідчить про те, що клієнт не залишив відгук. Заповнення таких пропусків середнім або іншим значенням могло б спотворити результати аналізу задоволеності клієнтів і призвести до хибних висновків.

---
## Завдання 9 (Середнє). Додавання обчислюваних колонок

Додайте до датафрейму нові колонки:
1. `month` — місяць замовлення (з колонки `date`).
2. `price_category` — категорія ціни: `'Низька'` (до 500), `'Середня'` (500–2000), `'Висока'` (понад 2000). Використайте `pd.cut()` або `np.select()`.
3. `discount_total` — сума зі знижкою 15%.

In [73]:
# Ваш код тут

df['month'] = df['date'].dt.month


df['price_category'] = np.select(conditions, choices, default='Невідомо')

df['discount_total'] = df['total'] * 0.85

print(df[['date', 'month', 'price', 'price_category', 'total', 'discount_total']])

          date  month    price price_category     total  discount_total
0   2023-01-01      1   963.38        Середня    963.38        818.8730
1   2023-01-02      1  1086.28        Середня   6517.68       5540.0280
2   2023-01-03      1  1883.84        Середня   7535.36       6405.0560
3   2023-01-04      1  2448.39         Висока   2448.39       2081.1315
4   2023-01-05      1  3110.36         Висока  21772.52      18506.6420
..         ...    ...      ...            ...       ...             ...
195 2023-07-15      7  2618.25         Висока  10473.00       8902.0500
196 2023-07-16      7  2976.81         Висока   2976.81       2530.2885
197 2023-07-17      7  2025.06         Висока  16200.48      13770.4080
198 2023-07-18      7   321.07         Низька    321.07        272.9095
199 2023-07-19      7  1709.23        Середня   5127.69       4358.5365

[200 rows x 6 columns]


In [72]:

df['price_category'] = pd.cut(
    df['price'],
    bins=[0, 500, 2000, float('inf')],
    labels=['Низька', 'Середня', 'Висока']
)

print(df['price_category'].value_counts())

price_category
Висока     121
Середня     70
Низька       9
Name: count, dtype: int64


**Висновок:** *(Який розподіл цінових категорій? Як метод pd.cut спрощує категоризацію?)*  

Після категоризації всі товари були розподілені на три цінові сегменти: «Низька», «Середня» та «Висока». Це дозволяє аналізувати продажі не лише за конкретною ціною, а й за ціновими групами, що значно спрощує сегментацію клієнтів і товарів.  

Метод `pd.cut()` автоматично розподіляє числові значення по заданих інтервалах, тому не потрібно вручну перевіряти кожну ціну та присвоювати їй категорію. Завдяки цьому дані стають більш зручними для аналізу, побудови звітів і порівняння результатів між різними ціновими сегментами.

---
## Завдання 10 (Середнє). Зведена таблиця (pivot table)

Побудуйте зведену таблицю за допомогою `pd.pivot_table()`, де:
- рядки — `city`,
- колонки — `category`,
- значення — сума `total`,
- агрегація — `sum`.

Додайте підсумки по рядках і колонках (`margins=True`).

In [74]:
# Ваш код тут
pivot = pd.pivot_table(
    df,
    values='total',
    index='city',
    columns='category',
    aggfunc='sum',
    margins=True
)

print(pivot)

category  Електроніка      Книги       Одяг   Продукти      Спорт         All
city                                                                         
Дніпро       65438.16  116118.97   31179.87  147336.56   72244.38   432317.94
Київ         58713.88   87339.03  103754.43  118283.10  113742.95   481833.39
Львів       154285.60  108959.19   74340.69   55768.80   43155.08   436509.36
Невідомо     21012.44   25984.45   46962.31   40508.03    5917.59   140384.82
Одеса        84611.65   26773.57   82361.08   60466.75   65539.34   319752.39
Харків       44068.06   15669.18  135537.23  161398.44  100203.22   456876.13
All         428129.79  380844.39  474135.61  583761.68  400802.56  2267674.03


 **Висновок:** *(Яке місто і яка категорія дають найбільший сумарний дохід? Що показують margins?)*  

 Найбільший сумарний дохід серед міст приносить Київ (≈481 833.39), що робить його ключовим регіоном за обсягом виручки. Серед товарних категорій найбільший внесок у загальний дохід має «Продукти» (≈583 761.68), що робить цю категорію головним драйвером продажів.  

З аналізу перетину видно, що окремі міста та категорії мають різну силу: наприклад, Львів сильно представлений в «Електроніці», а Харків — у «Продуктах», що свідчить про регіональні відмінності в попиті.  

`margins` дозволяє швидко побачити загальну картину бізнесу без додаткових обчислень:  

- які міста найбільш прибуткові,  
- які категорії формують основний дохід,  
- загальний оборот компанії.  


---
## Завдання 11 (Середнє). Підрахунок унікальних значень і частот

1. Підрахуйте кількість унікальних товарів (`product`) у кожній категорії.
2. Виведіть топ-5 найпопулярніших товарів за кількістю замовлень за допомогою `value_counts()`.
3. Визначте частку кожного способу оплати у відсотках (`normalize=True`).

In [ ]:
# Ваш код тут


**Висновок:** *(Який спосіб оплати найпопулярніший? Чи рівномірний розподіл між способами?)*

---
## Завдання 12 (Середнє). Аналіз повернень

Розрахуйте рівень повернень (частку повернених замовлень) для кожної категорії та кожного міста. Визначте:
- категорію з найвищим рівнем повернень,
- місто з найнижчим рівнем повернень.

Підказка: використайте `groupby()['is_returned'].mean()`.

In [ ]:
# Ваш код тут


**Висновок:** *(Яка категорія найчастіше повертається? Про що це може свідчити?)*

---
## Завдання 13 (Підвищена складність). Часовий аналіз

1. Встановіть колонку `date` як індекс датафрейму.
2. Розрахуйте щомісячний обсяг продажів (ресемплінг `resample('ME')` або `resample('M')`, sum по `total`).
3. Знайдіть найкращий і найгірший місяці за продажами.
4. Поверніть звичайний числовий індекс (`reset_index()`).

In [ ]:
# Ваш код тут


**Висновок:** *(Який місяць виявився найуспішнішим? Чи є сезонність у даних?)*

---
## Завдання 14 (Підвищена складність). Накопичувальні показники

1. Відсортуйте датафрейм за датою.
2. Додайте колонку `cumulative_total` — накопичувальний обсяг продажів з початку року (`cumsum()`).
3. Визначте, в який день було досягнуто 50% від загального річного обсягу продажів.

In [ ]:
# Ваш код тут


**Висновок:** *(Коли приблизно була зароблена половина річного доходу? Що це означає для бізнесу?)*

---
## Завдання 15 (Підвищена складність). apply() та lambda

1. Використайте `apply()` з lambda-функцією для створення колонки `total_with_tax`, де ПДВ становить 20%, але лише для замовлень на суму понад 1000 (решта — без змін).
2. Напишіть функцію `classify_rating(r)`, що повертає `'Поганий'` (1–2), `'Нейтральний'` (3), `'Хороший'` (4–5), `'Без оцінки'` (NaN). Застосуйте її до колонки `rating`.

In [ ]:
# Ваш код тут


**Висновок:** *(Яка частка замовлень має хороший рейтинг? Коли apply кращий за vectorized-операції?)*

---
## Завдання 16 (Підвищена складність). Злиття датафреймів (merge)

Створіть додатковий датафрейм `df_discounts` з колонками `category` та `discount_pct` (різні знижки для кожної категорії). Об'єднайте його з основним датафреймом за допомогою `pd.merge()` та розрахуйте `final_total` — суму після застосування відповідної знижки.

```python
df_discounts = pd.DataFrame({
    'category':     ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт'],
    'discount_pct': [0.10, 0.20, 0.05, 0.00, 0.15]
})
```

In [ ]:
df_discounts = pd.DataFrame({
    'category':     ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт'],
    'discount_pct': [0.10, 0.20, 0.05, 0.00, 0.15]
})

# Ваш код тут


**Висновок:** *(Яка категорія отримала найбільшу сумарну знижку? Коли варто використовувати merge замість map?)*

---
## Завдання 17 (Підвищена складність). Групові трансформації

1. Додайте колонку `city_avg_total` — середня сума замовлення по місту (для кожного рядка підставте середнє по його місту), використайте `groupby().transform('mean')`.
2. Додайте колонку `above_city_avg` — булева колонка, що показує, чи перевищує замовлення середнє по місту.
3. Підрахуйте частку таких замовлень для кожного міста.

In [ ]:
# Ваш код тут


**Висновок:** *(Чим transform відрізняється від звичайного groupby? У якому місті найбільша частка великих замовлень?)*

---
## Завдання 18 (Складне). Ковзне середнє та аномалії

1. Згрупуйте дані за місяцем та розрахуйте щоденну виручку.
2. Розрахуйте 7-денне ковзне середнє виручки (`rolling(7).mean()`).
3. Визначте дні-аномалії: дні, коли виручка відхилилася від ковзного середнього більш ніж на 2 стандартних відхилення.

In [ ]:
# Ваш код тут


**Висновок:** *(Скільки аномальних днів виявлено? Що може бути причиною різких відхилень виручки?)*

---
## Завдання 19 (Складне). Когортний аналіз

Виконайте спрощений когортний аналіз за місяцем першого замовлення:
1. Для кожного товару (`product`) знайдіть місяць першого продажу — це його «когорта».
2. Об'єднайте цю інформацію з основним датафреймом.
3. Побудуйте зведену таблицю: когорта (місяць першого продажу) проти місяця замовлення, значення — кількість замовлень.

Підказка: використайте `groupby('product')['date'].transform('min')` та `dt.to_period('M')`.

In [ ]:
# Ваш код тут


**Висновок:** *(Що показує когортна таблиця? Чи продовжують товари з першої когорти продаватися в наступних місяцях?)*

---
## Завдання 20 (Складне). Підсумковий звіт

Сформуйте підсумковий звіт у вигляді одного датафрейму `summary_report`, що містить для кожної категорії:
- загальний дохід (`total_revenue`),
- кількість замовлень (`order_count`),
- середній чек (`avg_order_value`),
- частку доходу від загального (`revenue_share_pct`),
- рівень повернень у відсотках (`return_rate_pct`),
- середній рейтинг без NaN (`avg_rating`).

Відсортуйте за `total_revenue` спадно. Збережіть результат у файл `summary_report.csv`.

In [ ]:
# Ваш код тут


**Висновок:** *(Яка категорія є лідером за доходом? Чи є зв'язок між рейтингом та рівнем повернень? Які рекомендації можна дати бізнесу на основі цього звіту?)*

---
## Загальні підсумки

Після виконання всіх завдань дайте відповідь на такі питання:

1. Які методи pandas виявилися для вас найбільш корисними?
2. Які труднощі виникли під час роботи з пропущеними значеннями?
3. Що можна додати до аналізу, щоб зробити його більш повним?

*(Ваші відповіді тут)*